src/volatile_memory.rs

```diff
@@ -618,14 +630,14 @@ impl Bytes<usize> for VolatileSlice<'_> {
         F: Read,
     {
         let end = self.compute_end_offset(addr, count)?;
-        unsafe {
-            // It is safe to overwrite the volatile memory. Accessing the guest
-            // memory as a mutable slice is OK because nothing assumes another
-            // thread won't change what is loaded.
-            let dst = &mut self.as_mut_slice()[addr..end];
-            src.read_exact(dst).map_err(Error::IOError)?;
-        }
-        Ok(())
+
+        // It is safe to overwrite the volatile memory. Accessing the guest memory as a mutable
+        // slice is OK because nothing assumes another thread won't change what is loaded. We also
+        // manually update the dirty bitmap below.
+        let dst = unsafe { &mut self.as_mut_slice()[addr..end] };
+
+        self.bitmap.mark_dirty(addr, count);
+        src.read_exact(dst).map_err(Error::IOError)
```

Comment:
In case the `read_exact` function fails here, we still mark the pages as dirty. Wouldn't it be best to update the bitmap only in case the write succeeds?
```suggestion
        src.read_exact(dst).map_err(Error::IOError)?;
        self.bitmap.mark_dirty(addr, count)
```

Reply:
I've added the `mark_dirty` first since even if the memory operation fails, I think we can still have partial writes to guest memory, and they are implicitly covered if the entire range was marked already. It is possible to generate false positives this way, but I think they are not significant errors in the context of tracking dirty memory areas.